File: src/main/linkedlist/cpp/ExecMemory.cpp
#include <iostream>
#include <fstream>
#include <string>
#include <vector>
#include <algorithm>
#include <chrono>
#include <stdexcept>
#include <filesystem>
#include <unistd.h>
#include "LinkedList.cpp"

// Inclui a biblioteca para contagem de alocação de memória
#include <malloc_count.h>
// Comando de compilação sugerido:
// g++ exec_memory.cpp ../../../../lib/malloc_count/malloc_count.c -o exec_memory.out -I../../../../lib/malloc_count

using namespace std;
namespace fs = std::filesystem;

/**
 * @brief Nomes dos métodos que serão testados para medição de memória
 *
 * Esta lista contém os nomes de todas as operações que serão executadas na fila
 * para medição do pico de memória durante cada operação.
 */
const vector<string> METHOD_NAMES = {
    "removeLast",
    "addLast",
    "getLast",
    "removeFirst",
    "addFirst",
    "getFirst",
    "removeByIndex",
    "add",
    "get"
};

/**
 * @brief Testa operações na fila medindo o pico de memória durante execução
 *
 * @param elements Vetor de inteiros contendo os elementos a serem inseridos na fila
 * @param writers Array de streams de arquivo para escrita dos resultados
 * @param currentLine Número da linha atual sendo processada (para feedback)
 * @param totalLines Total de linhas a serem processadas (para feedback)
 */
void testllOperations(const vector<int>& elements, ofstream writers[], int currentLine, int totalLines) {
    int length = elements.size();
    int middle = length / 2;
    const int RUNS = 5;

    // Matriz para armazenar os picos de memória de cada método em cada execução
    vector<vector<size_t>> memoryPeaks(METHOD_NAMES.size(), vector<size_t>(RUNS));

    // Cria e preenche a fila com os elementos fornecidos
    LinkedList ll = LinkedList();
    for (int num : elements) {
        ll.addLast(num);
    }

    for (int run = 0; run < RUNS; run++) {
        // Testa removeLast
        malloc_count_reset_peak();
        ll.removeLast();
        memoryPeaks[0][run] = malloc_count_peak() / 1024; // Convertendo para KB

        // Testa addLast
        malloc_count_reset_peak();
        ll.addLast(999);
        memoryPeaks[1][run] = malloc_count_peak() / 1024;

        // Testa getLast
        malloc_count_reset_peak();
        ll.getLast();
        memoryPeaks[2][run] = malloc_count_peak() / 1024;

        // Testa removeFirst
        malloc_count_reset_peak();
        ll.removeFirst();
        memoryPeaks[3][run] = malloc_count_peak() / 1024;

        // Testa addFirst
        malloc_count_reset_peak();
        ll.addFirst(999);
        memoryPeaks[4][run] = malloc_count_peak() / 1024;

        // Testa getFirst
        malloc_count_reset_peak();
        ll.getFirst();
        memoryPeaks[5][run] = malloc_count_peak() / 1024;

        // Testa remove (middle)
        malloc_count_reset_peak();
        ll.removeByIndex(middle);
        memoryPeaks[6][run] = malloc_count_peak() / 1024;

        // Testa add (middle)
        malloc_count_reset_peak();
        ll.add(999, middle);
        memoryPeaks[7][run] = malloc_count_peak() / 1024;
        
        // Testa get (middle)
        malloc_count_reset_peak();
        ll.get(middle);
        memoryPeaks[8][run] = malloc_count_peak() / 1024;
    }

    // Processa os resultados calculando a mediana dos picos de memória
    for (size_t method = 0; method < METHOD_NAMES.size(); method++) {
        sort(memoryPeaks[method].begin(), memoryPeaks[method].end());
        size_t medianPeak = memoryPeaks[method][RUNS / 2];
        
        // Escreve resultado no arquivo correspondente
        writers[method] << "linkedlist_cpp " << medianPeak << " " << length << endl;
        writers[method].flush();
    }

    // Exibe progresso para o usuário
    cerr << "\rProcessando linha " << currentLine << "/" << totalLines << flush;
}

/**
 * @brief Função principal que coordena a execução do programa
 *
 * @param argc Número de argumentos de linha de comando
 * @param argv Vetor de argumentos de linha de comando
 * @return int Código de saída (0 para sucesso, 1 para erro)
 */
int main(int argc, char* argv[]) {
    // Verifica argumentos de linha de comando
    if (argc < 2) {
        cerr << "ERRO: número de argumentos inválido!" << endl;
        return 1;
    }

    string inputFile = argv[1];
    string outputDir = "data/results/memory/";

    // Cria diretório de saída se não existir
    try {
        fs::create_directories(outputDir);
    } catch (const fs::filesystem_error& e) {
        cerr << "ERRO: Não foi possível criar o diretório de saída: "
             << outputDir << "\n" << e.what() << endl;
        return 1;
    }

    // Prepara arquivos de saída para cada método
    ofstream writers[METHOD_NAMES.size()];
    for (size_t i = 0; i < METHOD_NAMES.size(); i++) {
        string outputFile = outputDir + METHOD_NAMES[i] + ".data";
        bool fileExists = fs::exists(outputFile);
        bool fileIsEmpty = fileExists ? fs::file_size(outputFile) == 0 : true;

        writers[i].open(outputFile, ios::app);
        if (!writers[i]) {
            cerr << "\nErro ao abrir arquivo de saída: " << outputFile << endl;
            return 1;
        }

        // Escreve cabeçalho se o arquivo for novo ou estiver vazio
        if (!fileExists || fileIsEmpty) {
            writers[i] << "estrutura_linguagem memoria tamanho" << endl;
        }
    }

    try {
        // Abre arquivo de entrada
        ifstream reader(inputFile);
        if (!reader) {
            throw runtime_error("\nNão foi possível abrir o arquivo de entrada: " + inputFile);
        }

        // Conta total de linhas para feedback de progresso
        int totalLines = 0;
        string line;
        while (getline(reader, line)) {
            if (!line.empty()) totalLines++;
        }
        reader.clear();
        reader.seekg(0);

        // Processa cada linha do arquivo de entrada
        int currentLine = 0;
        while (getline(reader, line)) {
            if (line.empty()) continue;
            currentLine++;
            
            // Converte linha em vetor de inteiros
            vector<int> elements;
            size_t pos = 0;
            while ((pos = line.find(' ')) != string::npos) {
                string token = line.substr(0, pos);
                elements.push_back(stoi(token));
                line.erase(0, pos + 1);
            }
            if (!line.empty()) {
                elements.push_back(stoi(line));
            }

            // Testa operações com os elementos lidos
            testllOperations(elements, writers, currentLine, totalLines);
        }

        cerr << "\rConcluído! Processadas " << totalLines << " linhas." << endl;

        // Fecha todos os arquivos de saída
        for (size_t i = 0; i < METHOD_NAMES.size(); i++) {
            writers[i].close();
        }

    } catch (const exception& e) {
        // Tratamento de erros durante a execução
        cerr << "\nErro: " << e.what() << endl;
        for (size_t i = 0; i < METHOD_NAMES.size(); i++) {
            if (writers[i].is_open()) {
                writers[i].close();
            }
        }
        return 1;
    }

    return 0;
}

File: src/main/linkedlist/java/ExecMemory.java
import java.io.BufferedReader;
import java.io.File;
import java.io.FileReader;
import java.io.FileWriter;
import java.io.IOException;
import java.io.PrintWriter;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

// Inclui a biblioteca para contagem de alocação de memória
import org.github.jamm.MemoryMeter;
// Comando de compilação sugerido:
// javac -cp lib/jamm-0.4.0.jar src/main/linkedlist/java/*.java
// Comando de execução sugerido:
// java -cp lib/jamm-0.4.0.jar:src/main/linkedlist/java ExecMemory <arquivo_de_entrada>

/**
 * Classe para medição de consumo de memória de operações em uma estrutura ll.
 * Gera arquivos de saída com os picos de memória medidos para análise.
 */
public class ExecMemory {
    // Nomes das operações que serão medidas
    private static final String[] METHOD_NAMES = {
        "remove_last",
        "add_last",
        "get_last",
        "remove_first",
        "add_first",
        "get_first",
        "remove_middle",
        "add_middle",
        "get_middle"
    };

    // Número de execuções para medição (deve ser ímpar para cálculo da mediana)
    private static final int MEASUREMENT_RUNS = 5;
    // Diretório de saída
    private static final String OUTPUT_DIR = "data/results/memory/";
    // Instância do medidor de memória JAMM
    private static final MemoryMeter memoryMeter = MemoryMeter.builder().build();
    
    public static void main(String[] args) {
        if (args.length == 0) {
            System.err.println("ERRO: nenhum argumento!");
            System.exit(1);
        }

        // Arquivo de entrada com os dados de teste
        String inputFile = args[0];
        // Garante que o diretório de saída existe
        createOutputDirectory();

        try {
            // Lê todas as linhas do arquivo de entrada
            List<String> lines = readAllLines(inputFile);
            // Inicializa writers para os arquivos de saída
            PrintWriter[] writers = initializeWriters();

            // Processa cada linha e realiza as medições
            processLines(lines, writers);

            // Fecha todos os arquivos de saída
            closeWriters(writers);
            System.out.printf("\rFinalizado! Processadas %d linhas.%n", lines.size());
        } catch (IOException e) {
            System.err.println("ERRO: " + e.getMessage());
            e.printStackTrace();
            System.exit(1);
        }
    }

    private static void createOutputDirectory() {
        File outputDirFile = new File(OUTPUT_DIR);
        if (!outputDirFile.exists() && !outputDirFile.mkdirs()) {
            System.err.println("Falha ao criar diretório de saída: " + OUTPUT_DIR);
            System.exit(1);
        }
    }

    private static List<String> readAllLines(String inputFile) throws IOException {
        List<String> lines = new ArrayList<>();
        try (BufferedReader reader = new BufferedReader(new FileReader(inputFile))) {
            String line;
            while ((line = reader.readLine()) != null) {
                if (!line.isBlank()) {
                    lines.add(line);
                }
            }
        }
        return lines;
    }

    private static PrintWriter[] initializeWriters() throws IOException {
        PrintWriter[] writers = new PrintWriter[METHOD_NAMES.length];
        for (int i = 0; i < METHOD_NAMES.length; i++) {
            File outputFile = new File(OUTPUT_DIR + METHOD_NAMES[i] + ".data");
            boolean fileExists = outputFile.exists();
            boolean fileIsEmpty = fileExists ? outputFile.length() == 0 : true;

            // Abre arquivo em modo append (adiciona ao existente)
            writers[i] = new PrintWriter(new FileWriter(outputFile, true));

            // Se o arquivo não existia ou está vazio, escreve o cabeçalho
            if (!fileExists || fileIsEmpty) {
                writers[i].println("estrutura_linguagem memoria tamanho");
            }
        }
        return writers;
    }

    private static void processLines(List<String> lines, PrintWriter[] writers) {
        int totalLines = lines.size();
        for (int i = 0; i < totalLines; i++) {
            String line = lines.get(i);
            System.out.printf("\rProcessando linha %d/%d", i + 1, totalLines);
            System.out.flush();

            try {
                // Converte a linha em array de inteiros
                int[] elements = Arrays.stream(line.split("\\s+"))
                                     .mapToInt(Integer::parseInt)
                                     .toArray();
                // Realiza os testes com os elementos lidos
                testllOperations(writers, elements);
            } catch (NumberFormatException e) {
                System.err.printf("%nErro ao analisar linha %d: %s%n", i + 1, line);
            }
        }
    }

    private static void closeWriters(PrintWriter[] writers) {
        for (PrintWriter writer : writers) {
            if (writer != null) {
                writer.close();
            }
        }
    }
    
    private static void testllOperations(PrintWriter[] writers, int[] elements) {
        int length = elements.length;
        int middle = length / 2;
        long[][] memoryPeaks = new long[METHOD_NAMES.length][MEASUREMENT_RUNS];

        // Inicializa a fila com os elementos fornecidos
        LinkedList ll = new LinkedList();
        for (int num : elements) {
            ll.addLast(num);
        }

        // Fase de medição - registra os picos de memória
        for (int run = 0; run < MEASUREMENT_RUNS; run++) {
            performMemoryMeasurements(ll, middle, memoryPeaks, run);
        }

        // Calcula a mediana dos picos de memória e escreve nos arquivos
        for (int method = 0; method < METHOD_NAMES.length; method++) {
            Arrays.sort(memoryPeaks[method]);
            long medianMemory = memoryPeaks[method][MEASUREMENT_RUNS / 2] / 1024; // Convertendo para KB

            writers[method].printf("linkedlist_java %d %d%n", medianMemory, length);
            writers[method].flush();
        }
    }

    private static void performMemoryMeasurements(LinkedList ll, int middle, 
                                                long[][] memoryPeaks, int currentRun) {
        // Mede o pico de memória para cada operação
        memoryPeaks[0][currentRun] = measureMemoryPeak(ll, () -> ll.removeLast());
        memoryPeaks[1][currentRun] = measureMemoryPeak(ll, () -> ll.addLast(999));
        memoryPeaks[2][currentRun] = measureMemoryPeak(ll, () -> ll.getLast());
        memoryPeaks[3][currentRun] = measureMemoryPeak(ll, () -> ll.removeFirst());
        memoryPeaks[4][currentRun] = measureMemoryPeak(ll, () -> ll.addFirst(999));
        memoryPeaks[5][currentRun] = measureMemoryPeak(ll, () -> ll.getFirst());
        memoryPeaks[6][currentRun] = measureMemoryPeak(ll, () -> ll.removeByIndex(middle));
        memoryPeaks[7][currentRun] = measureMemoryPeak(ll, () -> ll.add(999, middle));
        memoryPeaks[8][currentRun] = measureMemoryPeak(ll, () -> ll.get(middle));
    }

    private static long measureMemoryPeak(LinkedList ll, Runnable operation) {
        // Executa a operação
        operation.run();
        
        // Retorna o consumo de memória do objeto ll
        return memoryMeter.measureDeep(ll);
    }
}

